<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>public · QGoF</title><meta name="title" content="public · QGoF"/><meta property="og:title" content="public · QGoF"/><meta property="twitter:title" content="public · QGoF"/><meta name="description" content="Documentation for QGoF."/><meta property="og:description" content="Documentation for QGoF."/><meta property="twitter:description" content="Documentation for QGoF."/><meta property="og:url" content="https://cecileane.github.io/QuartetNetworkGoodnessFit.jl/stable/lib/public/"/><meta property="twitter:url" content="https://cecileane.github.io/QuartetNetworkGoodnessFit.jl/stable/lib/public/"/><link rel="canonical" href="https://cecileane.github.io/QuartetNetworkGoodnessFit.jl/stable/lib/public/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="QGoF logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">QGoF</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">home</a></li><li><span class="tocitem">manual</span><ul><li><a class="tocitem" href="../../man/gof/">goodness of fit</a></li><li><a class="tocitem" href="../../man/simulate/">simulating quartet concordance factors</a></li><li><a class="tocitem" href="../../man/expected_qCFs/">expected concordance factors</a></li></ul></li><li><span class="tocitem">library</span><ul><li class="is-active"><a class="tocitem" href>public</a><ul class="internal"><li><a class="tocitem" href="#index"><span>index</span></a></li><li><a class="tocitem" href="#functions"><span>functions</span></a></li></ul></li><li><a class="tocitem" href="../internal/">internal</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">library</a></li><li class="is-active"><a href>public</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>public</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cecileane/QuartetNetworkGoodnessFit.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cecileane/QuartetNetworkGoodnessFit.jl/blob/master/docs/src/lib/public.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="public-documentation"><a class="docs-heading-anchor" href="#public-documentation">public documentation</a><a id="public-documentation-1"></a><a class="docs-heading-anchor-permalink" href="#public-documentation" title="Permalink"></a></h1><p>Documentation for <code>QuartetNetworkGoodnessFit</code>&#39;s public (exported) functions.</p><h2 id="index"><a class="docs-heading-anchor" href="#index">index</a><a id="index-1"></a><a class="docs-heading-anchor-permalink" href="#index" title="Permalink"></a></h2><ul><li><a href="#QuartetNetworkGoodnessFit.network_expectedCF-Tuple{HybridNetwork}"><code>QuartetNetworkGoodnessFit.network_expectedCF</code></a></li><li><a href="#QuartetNetworkGoodnessFit.quarnetGoFtest!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}"><code>QuartetNetworkGoodnessFit.quarnetGoFtest!</code></a></li><li><a href="#QuartetNetworkGoodnessFit.ticr!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}"><code>QuartetNetworkGoodnessFit.ticr!</code></a></li></ul><h2 id="functions"><a class="docs-heading-anchor" href="#functions">functions</a><a id="functions-1"></a><a class="docs-heading-anchor-permalink" href="#functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="QuartetNetworkGoodnessFit.network_expectedCF-Tuple{HybridNetwork}" href="#QuartetNetworkGoodnessFit.network_expectedCF-Tuple{HybridNetwork}"><code>QuartetNetworkGoodnessFit.network_expectedCF</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">network_expectedCF(net::HybridNetwork; showprogressbar=true,
        inheritancecorrelation=0)</code></pre><p>Calculate the quartet concordance factors (qCF) expected from the multispecies coalescent along network <code>net</code>. Output: <code>(q,t)</code> where <code>t</code> is a list of taxa, and <code>q</code> is a list of 4-taxon set objects of type <code>PhyloNetworks.QuartetT{datatype}</code>. In each element of <code>q</code>, <code>taxonnumber</code> gives the indices in <code>taxa</code> of the 4 taxa of interest; and <code>data</code> contains the 3 concordance factors, for the 3 unrooted topologies in the following order: <code>t1,t2|t3,t4</code>, <code>t1,t3|t2,t4</code> and <code>t1,t4|t2,t3</code>. This output is similar to that of <code>PhyloNetworks.countquartetsintrees</code> when 1 individual = 1 taxon, with 4-taxon sets listed in the same order (same output <code>t</code>, then same order of 4-taxon sets in <code>q</code>).</p><p>Assumption: the network should have <strong>edge lengths in coalescent units</strong>.</p><p>By default, lineages at a hybrid node come from a parent (chosen according to inheritance probabilities γ) <em>independently</em> across lineages. With option <code>inheritancecorrelation &gt; 0</code>, lineages have positive dependence, e.g. to model locus-specific inheritance probabilities, randomly drawn from a Beta distribution with mean γ across all loci. If <code>inheritancecorrelation</code> is set to 1, then all lineages at a given locus inherit from the same (randomly sampled) parent. More generally, the lineages&#39; parents are distributed according to a Dirichlet process with base distribution determined by the γ values, and with concentration parameter α = (1-r)/r, that is, r = 1/(1+α), where <code>r</code> is the input inheritance correlation.</p><p><strong>examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; using PhyloNetworks, QuartetNetworkGoodnessFit

julia&gt; # network with 3_2 cycles, causing some anomalous quartets
       net = readTopology(&quot;(D:1,((C:1,#H25:0):0.1,((((B1:10,B2:1):1.5,#H1:0):10.8,((A1:1,A2:1):0.001)#H1:0::0.5):0.5)#H25:0::0.501):1);&quot;);

julia&gt; # using PhyloPlots; plot(net, showedgelength=true);

julia&gt; q,t = network_expectedCF(net); # anomalous: A1, A2, {B1 or B2}, {C or D}
Calculation quartet CFs for 15 quartets...
0+---------------+100%
  ***************

julia&gt; show(q[1].taxonnumber)
[1, 2, 3, 4]
julia&gt; show(q[1].data)
[0.8885456713760765, 0.05572716431196175, 0.05572716431196175]

julia&gt; for qi in q
         println(join(t[qi.taxonnumber],&quot;,&quot;) * &quot;: &quot; * string(round.(qi.data, sigdigits=3)))
       end
A1,A2,B1,B2: [0.889, 0.0557, 0.0557]
A1,A2,B1,C: [0.168, 0.416, 0.416]
A1,A2,B2,C: [0.168, 0.416, 0.416]
A1,B1,B2,C: [0.0372, 0.0372, 0.926]
A2,B1,B2,C: [0.0372, 0.0372, 0.926]
A1,A2,B1,D: [0.168, 0.416, 0.416]
A1,A2,B2,D: [0.168, 0.416, 0.416]
A1,B1,B2,D: [0.0372, 0.0372, 0.926]
A2,B1,B2,D: [0.0372, 0.0372, 0.926]
A1,A2,C,D: [0.69, 0.155, 0.155]
A1,B1,C,D: [0.793, 0.103, 0.103]
A2,B1,C,D: [0.793, 0.103, 0.103]
A1,B2,C,D: [0.793, 0.103, 0.103]
A2,B2,C,D: [0.793, 0.103, 0.103]
B1,B2,C,D: [1.0, 9.42e-7, 9.42e-7]
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cecileane/QuartetNetworkGoodnessFit.jl/blob/ddeb3210fb90d24fed5ed4039d1e6b34efbe9115/src/quarnetconcordancefactors.jl#L1-L68">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="QuartetNetworkGoodnessFit.quarnetGoFtest!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}" href="#QuartetNetworkGoodnessFit.quarnetGoFtest!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}"><code>QuartetNetworkGoodnessFit.quarnetGoFtest!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">quarnetGoFtest!(net::HybridNetwork, df::DataFrame, optbl::Bool; quartetstat=:LRT, correction=:simulation, seed=1234, nsim=1000, verbose=false, keepfiles=false)
quarnetGoFtest!(net::HybridNetwork, dcf::DataCF,   optbl::Bool; kwargs...)</code></pre><p>Goodness-of-fit test for the adequacy of the multispecies network coalescent, to see if a given population or species network explains the quartet concordance factor data adequately. The network needs to be of level 1 at most (trees fullfil this condition), and have branch lengths in coalescent units. The test assumes a multinomial distribution for the observed quartet concordance factors (CF), such that information on the number of genes for each four-taxon set (<code>ngenes</code> field) must be present.</p><p>For each four-taxon set, an outlier p-value is obtained by comparing a test statistic (-2log likelihood ratio by default) to a chi-square distribution with 2 degrees of freedom (see below for other options).</p><p>The four-taxon sets are then categorized as outliers or not, according to their outlier p-values (outlier if p&lt;0.05). Finally, a one-sided goodness-of-fit test is performed on the frequency of outlier 4-taxon sets to see if there are more outliers than expected. The z-value for this test corresponds to the null hypothesis that 5% outlier p-values are &lt; 0.05 (versus more than 5%):</p><p class="math-container">\[z = \frac{\mathrm{proportion.outliers} - 0.05}{\sqrt{0.05 \times 0.95/\mathrm{number.4taxon.sets}}}.\]</p><p>This z-value corresponds to a test that assumes independent outlier p-values across 4-taxon sets: it makes no correction for dependence.</p><p>To correct for dependence with <code>correction=:simulation</code>, the distribution of z-values is obtained by simulating gene trees under the coalescent along the network (after branch length optimization if <code>optbl=true</code>) using <a href="https://github.com/cecileane/PhyloCoalSimulations.jl">PhyloCoalSimulations</a>. The z-score is calculated on each simulated data set. Under independence, these z-scores have mean 0 and variance 1. Under dependence, these z-scores still have mean 0, but an inflated variance. This variance σ² is estimated from the simulations, and the corrected p-value is obtained by comparing the original z value to N(0,σ²). When <code>correction=:none</code>, σ is taken to be 1 (independence): <em>not</em> recommended!</p><ul><li>The first version takes a <code>DataFrame</code> where each row corresponds to a given four-taxon set. The data frame is modified by having an additional another column containing the p-values corresponding to each four-taxon set.</li><li>The second version takes a <code>DataCF</code> object and modifies it by updating the expected concordance factors stored in that object.</li></ul><p>Note that <code>net</code> is <strong>not</strong> modified.</p><p><strong>arguments</strong></p><ul><li><code>optbl</code>: when <code>false</code>, branch lengths in <code>net</code> are taken as is, and need to be in coalescent units. When <code>optbl=true</code>, branch lengths in <code>net</code> are optimized, to optimize the pseudo log likelihood score as in SNaQ (see <a href="https://crsl4.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.topologyMaxQPseudolik!">here</a>). In both cases, any missing branch length is assigned a value with <a href="../internal/#QuartetNetworkGoodnessFit.ultrametrize!-Tuple{HybridNetwork, Bool}"><code>ultrametrize!</code></a>, which attempts to make the major tree ultrametric (but never modifies an existing edge length). Missing branch lengths may arise if they are not identifiable, such as lengths of external branches if there is a single allele per taxon. The network is returned as part of the output.</li></ul><p><strong>keyword arguments</strong></p><ul><li><code>quartetstat</code>: the test statistic used to obtain an outlier p-value for each four-taxon set, which is then compared to a chi-squared distribution with 2 degrees of freedom to get a p-value.<ul><li><code>:LRT</code> is the default, for the likelihood ratio: <span>$2n_\mathrm{genes} \sum_{j=1}^3 {\hat p}_j (\log{\hat p}_j - \log p_j)$</span> where <span>$p_j$</span> is the quartet CF expected from the network, and <span>${\hat p}_j$</span> is the quartet CF observed in the data.</li><li><code>:Qlog</code> for the Qlog statistics (Lorenzen, 1995): <span>$2n_\mathrm{genes} \sum_{j=1}^3 \frac{({\hat p}_j - p_j)^2}{p_j (\log{\hat p}_j - \log p_j)}$</span> and</li><li><code>:pearson</code> for Pearon&#39;s chi-squared statistic, which behaves poorly when one or more expected counts are low (e.g. less than 5): <span>$n_\mathrm{genes} \sum_{j=1}^3 \frac{({\hat p}_j - p_j)^2 }{p_j}$</span></li></ul></li><li><code>correction=:simulation</code> to correct for dependence across 4-taxon. Use <code>:none</code> to turn off simulations and the correction for dependence.</li><li><code>seed=1234</code>: master seed to control the seeds for gene tree simulations.</li><li><code>nsim=1000</code>: number of simulated data sets. Each data set is simulated to have the median number of genes that each 4-taxon sets has data for.</li><li><code>verbose=false</code>: turn to <code>true</code> to see progress of simulations and to diagnose potential issues.</li><li><code>keepfiles=false</code>: if true, simulated gene trees are written to files, one file for each of the 1000 simulations. If created (with <code>keepfiles=true</code>), these files are stored in a newly created folder whose name starts with <code>jl_</code> and is placed in the current directory.</li></ul><p><strong>output</strong></p><ol><li>p-value of the overall goodness-of-fit test (corrected for dependence if requested)</li><li>uncorrected z value test statistic</li><li>estimated σ for the test statistic used for the correction (1.0 if no correction)</li><li>a vector of outlier p-values, one for each four-taxon set</li><li>network (first and second versions): <code>net</code> with loglik field updated if <code>optbl</code> is false;  copy of <code>net</code> with optimized branch lengths and loglik if <code>optbl</code> is true</li><li>in case <code>correction = :simulation</code>, vector of simulated z values (<code>nothing</code> if <code>correction = :none</code>). These z-values could be used to calculate an empirical p-value (instead of the p-value in #1), as the proportion of simulated z-values that are ⩾ the observed z-value (in #2).</li></ol><p><strong>references</strong></p><ul><li><p>Ruoyi Cai &amp; Cécile Ané (2021). Assessing the fit of the multi-species network coalescent to multi-locus data. Bioinformatics, 37(5):634-641. doi: <a href="https://doi.org/10.1093/bioinformatics/btaa863">10.1093/bioinformatics/btaa863</a></p></li><li><p>Lorenzen (1995). A new family of goodness-of-fit statistics for discrete multivariate data. Statistics &amp; Probability Letters, 25(4):301-307. doi: <a href="https://doi.org/10.1016/0167-7152(94)00234-8">10.1016/0167-7152(94)00234-8</a></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cecileane/QuartetNetworkGoodnessFit.jl/blob/ddeb3210fb90d24fed5ed4039d1e6b34efbe9115/src/quarnetGoF.jl#L1-L118">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="QuartetNetworkGoodnessFit.ticr!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}" href="#QuartetNetworkGoodnessFit.ticr!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}"><code>QuartetNetworkGoodnessFit.ticr!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ticr!(net::HybridNetwork, df::DataFrame, optbl::Bool; quartetstat, test)
ticr!(net::HybridNetwork, dcf::DataCF,   optbl::Bool; quartetstat, test)
ticr(dcf::DataCF, quartetstat::Symbol, test::Symbol)</code></pre><p>Goodness-of-fit test for the adequacy of the multispecies network coalescent, to see if a given population or species network explains the quartet concordance factor data adequately. See Stenz et al 2015 and <a href="http://www.stat.wisc.edu/~ane/publis/2015Stenz_TICR_addendum.pdf">addendum</a> for the method on trees, from which the acronym TICR was derived: Tree Incongruence Checking with R.</p><p>The tree / network needs to have branch lengths in coalescent units, and must be of level 1. It must be fully resolved, such that the &quot;major&quot; quartet is clearly defined, even though branch lengths can be 0.</p><p>The Dirichlet distribution is used to fit the distribution of the observed quartet concordance factors (CF), with a concentration parameter estimated from the data.</p><p>The four-taxon sets are then binned into categories according to their outlier p-values: <code>0-0.01</code>, <code>0.01-0.05</code>, <code>0.05-0.10</code>, and <code>0.10-1</code>. Finally, a one-sided goodness-of-fit test is performed on these binned frequencies to determine if they depart from an expected 5% p-values being below 0.05 (versus more than 5%), using the default <code>test = :onesided</code>. Alternatively, the option <code>test = :goodness</code> carries out the overall goodness-of-fit chi-square test proposed by Stenz et al. (2015), to test for any kind of departure from the expected frequencies (0.01, 0.04, 0.05, 0.90) across all 4 bins.</p><ul><li>The first version takes a <code>DataFrame</code> where each row corresponds to a given four-taxon set. The data frame is modified by having an additional another column containing the outlier p-values corresponding to each four-taxon set.</li><li>The second version takes a <code>DataCF</code> object and modifies it by updating the expected concordance factors stored in that object.</li><li>The last version (which all others call) assumes that the expected concordance factors in the <code>DataCF</code> object are correctly calculated from the test network.</li></ul><p><strong>arguments</strong></p><ul><li><code>optbl</code>: when <code>false</code>, the loglik field of <code>net</code> is updated but branch lengths are taken as is. When <code>true</code>, a copy of <code>net</code> is used to conduce the test, with updated branch lengths (in coalescent units) and updated loglik. This network is returned.</li><li><code>quartetstat = :maxCF</code>: test statistic used to obtain an outlier p-value for each four-taxon set. By default, it is the absolute difference between the observed CF and expected CF of the major resolution (which has the largest CF) if <code>quartetstat=:maxCF</code>, as used in Stenz et al. (2015). The other option is <code>quartetstat=:minpval</code>, in which case the absolute difference between the observed CF and expected CF is calculated for all 3 resolutions, leading to 3 (non-independent) p-values. The outlier p-value for a given four-taxon set is taken as the minimum of these 3 p-values. This option can detect all kinds of departure from the network model for a given four-taxon set, but is not recommended because it is very liberal.</li><li><code>test = :onesided</code>: the overall goodness-of-fit test performed on binned frequencies of quartet outlier p-values; see above.</li></ul><p><strong>output</strong></p><ol><li>p-value of the overall goodness-of-fit test</li><li>test statistic (z value)</li><li>a dictionary of the count of p-values in each of the four category</li><li>a vector of two test parameters for the Dirichlet distribution:<ul><li>value of the concentration parameter α</li><li>value of the pseudo likelihood (optimized at α)</li></ul></li><li>a vector of outlier p-values, one for each four-taxon set</li><li>network (first and second versions): <code>net</code> with loglik field updated if <code>optbl</code> is false;  copy of <code>net</code> with optimized branch lengths and loglik if <code>optbl</code> is true</li></ol><p><strong>references</strong></p><p>NWM Stenz, B Larget, DA Baum and C Ané (2015). Exploring tree-like and non-tree-like patterns using genome sequences: An example using the inbreeding plant species <em>Arabidopsis thaliana</em> (L.) Heynh. Systematic Biology, 64(5):809-823. doi: <a href="https://doi.org/10.1093/sysbio/syv039">10.1093/sysbio/syv039</a></p><p>see also this <a href="http://www.stat.wisc.edu/~ane/publis/2015Stenz_TICR_addendum.pdf">addendum</a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cecileane/QuartetNetworkGoodnessFit.jl/blob/ddeb3210fb90d24fed5ed4039d1e6b34efbe9115/src/ticr.jl#L1-L81">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../man/expected_qCFs/">« expected concordance factors</a><a class="docs-footer-nextpage" href="../internal/">internal »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Monday 20 May 2024 18:49">Monday 20 May 2024</span>. Using Julia version 1.10.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
