var documenterSearchIndex = {"docs":
[{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"using PhyloNetworks, QuartetNetworkGoodnessFit\nENV[\"COLUMNS\"] = 125 # displaysize(stdout)[2] is 80 by default","category":"page"},{"location":"man/expected_qCFs/#expected-concordance-factors","page":"expected concordance factors","title":"expected concordance factors","text":"","category":"section"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"The quartet concordance factors expected from a network can be calculated exactly on a network of any level. This can be done here using a recursive algorithm – although it may run slowly as its complexity was not optimized. Below is an example, with a network on 6 taxa with 2 reticulations.","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"net = readTopology(\"(D:1,((C:1,#H25:0):0.1,((((B1:10,B2:1):1.5,#H1:0):10.8,((A1:1,A2:1):0.001)#H1:0::0.5):0.5)#H25:0::0.501):1);\");\neCFs,t = network_expectedCF(net);\nt # taxon list\nfirst(eCFs, 2) # first 2 4-taxon sets, each with 3 quartet CFs. taxon numbers are indices in the taxon list above","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"To look at these quartet CFs, we define below a function to convert them to a string. This string can then be printed to the screen (shown below with print) or written to a file (using write, not shown).","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"qCFstring(qlist,taxa, sigdigits=4) = join(\n  [join(taxa[q.taxonnumber],\",\") * \": \" * string(round.(q.data, sigdigits=sigdigits))\n   for q in qlist],   \"\\n\");\nprint(qCFstring(eCFs,t))","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"For each 4-taxon set above, the 3 concordance factors are for the quartets listed in this order: 12|34, 13|24, 14|23 where 1,2,3,4 refer to the order of the taxa listed to the left.","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"We can also convert our list to a data frame:","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"using DataFrames\ndf = DataFrame(\n    t1 = [t[q.taxonnumber[1]] for q in eCFs],\n    t2 = [t[q.taxonnumber[2]] for q in eCFs],\n    t3 = [t[q.taxonnumber[3]] for q in eCFs],\n    t4 = [t[q.taxonnumber[4]] for q in eCFs],\n    CF12_34 = [q.data[1] for q in eCFs],\n    CF13_24 = [q.data[2] for q in eCFs],\n    CF14_23 = [q.data[3] for q in eCFs],\n)","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"If we wanted to compare with the observed frequency among 100 gene trees, we could use data frame manipulations to join the two data frames:","category":"page"},{"location":"man/expected_qCFs/","page":"expected concordance factors","title":"expected concordance factors","text":"using PhyloCoalSimulations\ngenetrees = simulatecoalescent(net, 100, 1); # 1000 genes, 1 individual / pop\ndf_sim = writeTableCF(countquartetsintrees(genetrees; showprogressbar=false)...);\nfirst(df_sim, 2)\nselect!(df_sim, Not(:ngenes)); # delete column with number of genes\ndf_both = outerjoin(df, df_sim; on=[:t1,:t2,:t3,:t4], renamecols = \"exact\" => \"sim\")","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"using QuartetNetworkGoodnessFit\nENV[\"COLUMNS\"] = 85 # displaysize(stdout)[2] is 80 by default","category":"page"},{"location":"man/gof/#goodness_of_fit_1","page":"goodness of fit","title":"goodness of fit of a candidate network","text":"","category":"section"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"The example data set in this package is very small (5 taxa) on purpose: to make this tutorial run quickly. The data list all four-taxon sets, and the concordance factor (CF) for each of the 3 topologies on these 4 taxa, that is, the proportion of genes estimated to have each 4-taxon unrooted topology.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"using QuartetNetworkGoodnessFit, DataFrames, CSV\nqCF = DataFrame(CSV.File(joinpath(dirname(pathof(QuartetNetworkGoodnessFit)), \"..\",\"test\",\"example_qCF_5taxa.csv\")), copycols=false);\nqCF","category":"page"},{"location":"man/gof/#testing-candidate-networks","page":"goodness of fit","title":"testing candidate networks","text":"","category":"section"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Let's say we have two candidate networks to test: one that is just a tree (net0), and the other that has one reticulation (net1).","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"using PhyloNetworks\nnet1 = readTopology(\"((((D,C),((A,B))#H1),(#H1,E)),O);\");\nnet0 = readTopology(\"((((D,C),(A,B)),E),O);\"); # also: majorTree(net1)","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"net0 has clades AB and CD sister to each other, with E sister to ABCD and an outgroup O. net1 has net0 as its major tree, and has a reticulation going from E into the stem lineage to AB.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"The goal is to test the goodness-of-fit of each topology.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"It's best to start with a small number of simulations for the test, to check that there are no issue. Then we can re-run with a larger number. Note that the first use of a function includes time to compile the function, so it takes longer than the second use. Below, the option true means that we want to optimize the branch lengths in the network, in coalescent units, before quantifying the goodness-of-fit.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res0 = quarnetGoFtest!(net0, qCF, true; seed=201, nsim=3);\nnothing # hide","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Before re-running with more simulations, we can update the network net0 to the version that has optimized branch lengths:","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"net0 = res0[5]","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Now we re-run the test using the option false to not re-optimize branch lengths. We use nsim=200 simulations below to make this example faster. For a real data analysis, delete the nsim option to use the default instead (1000) or specify a higher value.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res0 = quarnetGoFtest!(net0, qCF, false; seed=234, nsim=200);","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"In the result, the first 3 numbers are the p-value of the overall goodness-of-fit test, the uncorrected z-value, and the estimate of σ to correct the z-value for dependence:","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res0[[1,2,3]] # p-value, uncorrected z, σ","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Here, we have evidence that the tree net0 does not fit the data adequately. Note that σ=1 corresponds to independent quartet outlier p-values. Typical estimates of σ are quite larger than 1, increase with the number of taxa, and increase with longer branch lengths (for a given number of taxa).","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"The next element is the list of outlier p-values, which was also added to the data frame that contains our data:","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res0[4]\nqCF[:,[:t1,:t2,:t3,:t4,:p_value]]","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"The very small overall p-value indicated an excess of outlier four-taxon sets. Here we see what these outliers are: all four-taxon sets containing O, E and either A or B are strong outliers (very small outlier p-values).","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"In fact, the CF data were simulated on net1, in which the AB clade received gene flow from E. We can re-run an analysis using net1 this time:","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res1 = quarnetGoFtest!(net1, qCF, true; seed=271, nsim=1000);\nres1[[1,2,3]] # p-value, uncorrected z, σ","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"This network is found to provide an adequate absolute fit (as expected from how the data were simulated).","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Note that after optimization of branch lengths and γs to best fit the CF data, the network was \"ultrametrized\" along its major tree to assign values to missing edge lengths. External edges are typically missing a length in coalescent units if there was a single individual sampled per species, for example.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"res1[5]","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"For more options, see quarnetGoFtest!, such as for the outlier test statistic (G or likelihood ratio test by default).","category":"page"},{"location":"man/gof/#parallel-computations","page":"goodness of fit","title":"parallel computations","text":"","category":"section"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"For larger networks, the large number of four-taxon sets causes the test to run more slowly. The computations can be parallelized by providing more processors to Julia. For instance, this can be done by starting julia with the -p option:","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"julia -p 3 # 3 worker processors, 4 processors total","category":"page"},{"location":"man/gof/#empirical-p-value","page":"goodness of fit","title":"empirical p-value","text":"","category":"section"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"\"Nice\" networks with long branch lengths in coalescent units lead to \"nice\" expected quartet concordance factors close to [100%, 0%, 0%], when gene trees are expected to agree on a good proportion of 4-taxon sets. On these networks and data sets with few loci, the distribution of the test z-value may be too far from a normal approximation. quarnetGoFtest! attempts to detect this issue. If it does, it sends a warning suggesting to use an empirical p-value instead of using the p-value returned by the first element of the output (e.g. res1[1]).","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"Here is example code to calculate the p-value from the empirical distribution of simulated z-values. Note that a large number of simulations is needed for this. We used nsim=1000 to test network 1 above, which is a good place to start.","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"zvalue_observed = res1[2]\nzvalue_bootstrap = sort!(res1[6]) # sorted z-values simulated under the network\nlength(zvalue_bootstrap) # just to check we ran many simulations\nusing Statistics # to access \"mean\"\npvalue = mean(zvalue_bootstrap .>= zvalue_observed) # one-sided test: Prob(Z > z)","category":"page"},{"location":"man/gof/","page":"goodness of fit","title":"goodness of fit","text":"In this case we get a p-value of 1.0, which is consisten with all individual outlier p-values being large, for all four-taxon sets as seen earlier: the quartet CFs couldn't fit net1 better.","category":"page"},{"location":"man/simulate/#quartet-concordance-factor-simulation","page":"simulating quartet concordance factors","title":"quartet concordance factor simulation","text":"","category":"section"},{"location":"man/simulate/#using-PhyloCoalSimulations","page":"simulating quartet concordance factors","title":"using PhyloCoalSimulations","text":"","category":"section"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"The correction for dependence across 4-taxon uses a simulation approach. We provide more examples here to simulate gene trees and extract the quartet concordance factors observed in these simulated gene trees.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"We use PhyloCoalSimulations for the simulation of trees under the network multispecies coalescent, assuming that the network edges lengths are in coalescent units.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"using PhyloNetworks, QuartetNetworkGoodnessFit, PhyloCoalSimulations\nnet1 = readTopology(\"((((D:0.2,C:0.2):2.4,((A:0.4,B:0.4):1.1)#H1:1.1::0.7):2.0,(#H1:0.0::0.3,E:1.5):3.1):1.0,O:5.6);\");\nngenes = 30\ngenetrees = simulatecoalescent(net1, ngenes, 1); # 1 individual / species\nobsCF, t = countquartetsintrees(genetrees; showprogressbar=true);\ndf = writeTableCF(obsCF, t)","category":"page"},{"location":"man/simulate/#using-Hybrid-Lambda","page":"simulating quartet concordance factors","title":"using Hybrid-Lambda","text":"","category":"section"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"warning: Warning\nHybrid-Lambda is not recommended (see Allman, Baños & Rhodes 2022) but this section may still be of interest to some and is left for completeness.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"The example qCF data used earlier were originally simulated as shown below using hybrid-Lambda, which comes with version v0.3 of QuartetNetworkGoodnessFit but not with later versions.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"In the code below, we show how to call hybrid-lambda within julia, asking for the simulation of 200 genes (-num 200). The major issue that this code solves is that hybrid-lambda does not use the standard extended Newick format, but we show code to generate the format expected by hybrid-lambda, from the standard format.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"Note that hybrid-lambda assumes that the tree or network is ultrametric, but runs even if this assumption is not met.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"If you required QuartetNetworkGoodnessFit v0.3, then do this:","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"hl = QuartetNetworkGoodnessFit.hybridlambda # path to hybrid-lambda simulator, on local machine","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"otherwise, install Hybrid-Lambda separately, then define hl as a string that gives the path to the Hybrid-Lambda executable on your local machine. Below, we use Julia to convert the standard Newick network into the format used by Hybrid-Lambda, run it, read the gene trees from the file created by Hybrid-Lambda, and finally calculate quartet concordance factors from these gene trees.","category":"page"},{"location":"man/simulate/","page":"simulating quartet concordance factors","title":"simulating quartet concordance factors","text":"net1 = readTopology(\"((((D:0.2,C:0.2):2.4,((A:0.4,B:0.4):1.1)#H1:1.1::0.7):2.0,(#H1:0.0::0.3,E:1.5):3.1):1.0,O:5.6);\");\nnet1HL = hybridlambdaformat(net1) # format for the network, to use below by hybrid-lambda\nrun(`$hl -spcu \"((((D:0.2,C:0.2)I1:2.4,((A:0.4,B:0.4)I2:1.1)H1#0.7:1.1)I3:2.0,(H1#0.7:0.0,E:1.5)I4:3.1)I5:1.0,O:5.6)I6;\" -num 200 -seed 123 -o \"genetrees\"`)\ntreelist = readMultiTopology(\"genetrees_coal_unit\")\nobsCF = writeTableCF(countquartetsintrees(treelist)...)","category":"page"},{"location":"lib/public/#public-documentation","page":"public","title":"public documentation","text":"","category":"section"},{"location":"lib/public/","page":"public","title":"public","text":"Documentation for QuartetNetworkGoodnessFit's public (exported) functions.","category":"page"},{"location":"lib/public/#index","page":"public","title":"index","text":"","category":"section"},{"location":"lib/public/","page":"public","title":"public","text":"Pages = [\"public.md\"]","category":"page"},{"location":"lib/public/#functions","page":"public","title":"functions","text":"","category":"section"},{"location":"lib/public/","page":"public","title":"public","text":"Modules = [QuartetNetworkGoodnessFit]\nPrivate = false\nOrder   = [:function]","category":"page"},{"location":"lib/public/#QuartetNetworkGoodnessFit.network_expectedCF-Tuple{HybridNetwork}","page":"public","title":"QuartetNetworkGoodnessFit.network_expectedCF","text":"network_expectedCF(net::HybridNetwork; showprogressbar=true,\n        inheritancecorrelation=0)\n\nCalculate the quartet concordance factors (qCF) expected from the multispecies coalescent along network net. Output: (q,t) where t is a list of taxa, and q is a list of 4-taxon set objects of type PhyloNetworks.QuartetT{datatype}. In each element of q, taxonnumber gives the indices in taxa of the 4 taxa of interest; and data contains the 3 concordance factors, for the 3 unrooted topologies in the following order: t1,t2|t3,t4, t1,t3|t2,t4 and t1,t4|t2,t3. This output is similar to that of PhyloNetworks.countquartetsintrees when 1 individual = 1 taxon, with 4-taxon sets listed in the same order (same output t, then same order of 4-taxon sets in q).\n\nAssumption: the network should have edge lengths in coalescent units.\n\nBy default, lineages at a hybrid node come from a parent (chosen according to inheritance probabilities γ) independently across lineages. With option inheritancecorrelation > 0, lineages have positive dependence, e.g. to model locus-specific inheritance probabilities, randomly drawn from a Beta distribution with mean γ across all loci. If inheritancecorrelation is set to 1, then all lineages at a given locus inherit from the same (randomly sampled) parent. More generally, the lineages' parents are distributed according to a Dirichlet process with base distribution determined by the γ values, and with concentration parameter α = (1-r)/r, that is, r = 1/(1+α), where r is the input inheritance correlation.\n\nexamples\n\njulia> using PhyloNetworks, QuartetNetworkGoodnessFit\n\njulia> # network with 3_2 cycles, causing some anomalous quartets\n       net = readTopology(\"(D:1,((C:1,#H25:0):0.1,((((B1:10,B2:1):1.5,#H1:0):10.8,((A1:1,A2:1):0.001)#H1:0::0.5):0.5)#H25:0::0.501):1);\");\n\njulia> # using PhyloPlots; plot(net, showedgelength=true);\n\njulia> q,t = network_expectedCF(net); # anomalous: A1, A2, {B1 or B2}, {C or D}\nCalculation quartet CFs for 15 quartets...\n0+---------------+100%\n  ***************\n\njulia> show(q[1].taxonnumber)\n[1, 2, 3, 4]\njulia> show(q[1].data)\n[0.8885456713760765, 0.05572716431196175, 0.05572716431196175]\n\njulia> for qi in q\n         println(join(t[qi.taxonnumber],\",\") * \": \" * string(round.(qi.data, sigdigits=3)))\n       end\nA1,A2,B1,B2: [0.889, 0.0557, 0.0557]\nA1,A2,B1,C: [0.168, 0.416, 0.416]\nA1,A2,B2,C: [0.168, 0.416, 0.416]\nA1,B1,B2,C: [0.0372, 0.0372, 0.926]\nA2,B1,B2,C: [0.0372, 0.0372, 0.926]\nA1,A2,B1,D: [0.168, 0.416, 0.416]\nA1,A2,B2,D: [0.168, 0.416, 0.416]\nA1,B1,B2,D: [0.0372, 0.0372, 0.926]\nA2,B1,B2,D: [0.0372, 0.0372, 0.926]\nA1,A2,C,D: [0.69, 0.155, 0.155]\nA1,B1,C,D: [0.793, 0.103, 0.103]\nA2,B1,C,D: [0.793, 0.103, 0.103]\nA1,B2,C,D: [0.793, 0.103, 0.103]\nA2,B2,C,D: [0.793, 0.103, 0.103]\nB1,B2,C,D: [1.0, 9.42e-7, 9.42e-7]\n\n\n\n\n\n\n","category":"method"},{"location":"lib/public/#QuartetNetworkGoodnessFit.quarnetGoFtest!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}","page":"public","title":"QuartetNetworkGoodnessFit.quarnetGoFtest!","text":"quarnetGoFtest!(net::HybridNetwork, df::DataFrame, optbl::Bool; quartetstat=:LRT, correction=:simulation, seed=1234, nsim=1000, verbose=false, keepfiles=false)\nquarnetGoFtest!(net::HybridNetwork, dcf::DataCF,   optbl::Bool; kwargs...)\n\nGoodness-of-fit test for the adequacy of the multispecies network coalescent, to see if a given population or species network explains the quartet concordance factor data adequately. The network needs to be of level 1 at most (trees fullfil this condition), and have branch lengths in coalescent units. The test assumes a multinomial distribution for the observed quartet concordance factors (CF), such that information on the number of genes for each four-taxon set (ngenes field) must be present.\n\nFor each four-taxon set, an outlier p-value is obtained by comparing a test statistic (-2log likelihood ratio by default) to a chi-square distribution with 2 degrees of freedom (see below for other options).\n\nThe four-taxon sets are then categorized as outliers or not, according to their outlier p-values (outlier if p<0.05). Finally, a one-sided goodness-of-fit test is performed on the frequency of outlier 4-taxon sets to see if there are more outliers than expected. The z-value for this test corresponds to the null hypothesis that 5% outlier p-values are < 0.05 (versus more than 5%):\n\nz = fracmathrmproportionoutliers - 005sqrt005 times 095mathrmnumber4taxonsets\n\nThis z-value corresponds to a test that assumes independent outlier p-values across 4-taxon sets: it makes no correction for dependence.\n\nTo correct for dependence with correction=:simulation, the distribution of z-values is obtained by simulating gene trees under the coalescent along the network (after branch length optimization if optbl=true) using PhyloCoalSimulations. The z-score is calculated on each simulated data set. Under independence, these z-scores have mean 0 and variance 1. Under dependence, these z-scores still have mean 0, but an inflated variance. This variance σ² is estimated from the simulations, and the corrected p-value is obtained by comparing the original z value to N(0,σ²). When correction=:none, σ is taken to be 1 (independence): not recommended!\n\nThe first version takes a DataFrame where each row corresponds to a given four-taxon set. The data frame is modified by having an additional another column containing the p-values corresponding to each four-taxon set.\nThe second version takes a DataCF object and modifies it by updating the expected concordance factors stored in that object.\n\nNote that net is not modified.\n\narguments\n\noptbl: when false, branch lengths in net are taken as is, and need to be in coalescent units. When optbl=true, branch lengths in net are optimized, to optimize the pseudo log likelihood score as in SNaQ (see here). In both cases, any missing branch length is assigned a value with ultrametrize!, which attempts to make the major tree ultrametric (but never modifies an existing edge length). Missing branch lengths may arise if they are not identifiable, such as lengths of external branches if there is a single allele per taxon. The network is returned as part of the output.\n\nkeyword arguments\n\nquartetstat: the test statistic used to obtain an outlier p-value for each four-taxon set, which is then compared to a chi-squared distribution with 2 degrees of freedom to get a p-value. The default is :LRT for the likelihood ratio: 2n_mathrmgenes sum_j=1^3 hat p_j (loghat p_j - log p_j) where p_j is the quartet CF expected from the network, and hat p_j is the quartet CF observed in the data.   Alternatives are :Qlog for the Qlog statistics (Lorenzen, 1995): 2n_mathrmgenes sum_j=1^3 frac(hat p_j - p_j)^2p_j (loghat p_j - log p_j)   and :pearson for Pearon's chi-squared statistic, which behaves poorly when the expected count is low (e.g. less than 5): n_mathrmgenes sum_j=1^3 frac(hat p_j - p_j)^2 p_j\ncorrection=:simulation to correct for dependence across 4-taxon. Use :none to turn off simulations and the correction for dependence.\nseed=1234: master seed to control the seeds for gene tree simulations.\nnsim=1000: number of simulated data sets. Each data set is simulated to have the median number of genes that each 4-taxon sets has data for.\nverbose=false: turn to true to see progress of simulations and to diagnose potential issues.\nkeepfiles=false: if true, simulated gene trees are written to files, one file for each of the 1000 simulations. If created (with keepfiles=true), these files are stored in a newly created folder whose name starts with jl_ and is placed in the current directory.\n\noutput\n\np-value of the overall goodness-of-fit test (corrected for dependence if requested)\nuncorrected z value test statistic\nestimated σ for the test statistic used for the correction (1.0 if no correction)\na vector of outlier p-values, one for each four-taxon set\nnetwork (first and second versions): net with loglik field updated if optbl is false;  copy of net with optimized branch lengths and loglik if optbl is true\nin case correction = :simulation, vector of simulated z values (nothing if correction = :none). These z-values could be used to calculate an empirical p-value (instead of the p-value in #1), as the proportion of simulated z-values that are ⩾ the observed z-value (in #2).\n\nreferences\n\nRuoyi Cai & Cécile Ané (2021). Assessing the fit of the multi-species network coalescent to multi-locus data. Bioinformatics, 37(5):634-641. doi: 10.1093/bioinformatics/btaa863\nLorenzen (1995). A new family of goodness-of-fit statistics for discrete multivariate data. Statistics & Probability Letters, 25(4):301-307. doi: 10.1016/0167-7152(94)00234-8\n\n\n\n\n\n","category":"method"},{"location":"lib/public/#QuartetNetworkGoodnessFit.ticr!-Tuple{HybridNetwork, DataFrames.DataFrame, Bool}","page":"public","title":"QuartetNetworkGoodnessFit.ticr!","text":"ticr!(net::HybridNetwork, df::DataFrame, optbl::Bool; quartetstat, test)\nticr!(net::HybridNetwork, dcf::DataCF,   optbl::Bool; quartetstat, test)\nticr(dcf::DataCF, quartetstat::Symbol, test::Symbol)\n\nGoodness-of-fit test for the adequacy of the multispecies network coalescent, to see if a given population or species network explains the quartet concordance factor data adequately. See Stenz et al 2015 and addendum for the method on trees, from which the acronym TICR was derived: Tree Incongruence Checking with R.\n\nThe tree / network needs to have branch lengths in coalescent units, and must be of level 1. It must be fully resolved, such that the \"major\" quartet is clearly defined, even though branch lengths can be 0.\n\nThe Dirichlet distribution is used to fit the distribution of the observed quartet concordance factors (CF), with a concentration parameter estimated from the data.\n\nThe four-taxon sets are then binned into categories according to their outlier p-values: 0-0.01, 0.01-0.05, 0.05-0.10, and 0.10-1. Finally, a one-sided goodness-of-fit test is performed on these binned frequencies to determine if they depart from an expected 5% p-values being below 0.05 (versus more than 5%), using the default test = :onesided. Alternatively, the option test = :goodness carries out the overall goodness-of-fit chi-square test proposed by Stenz et al. (2015), to test for any kind of departure from the expected frequencies (0.01, 0.04, 0.05, 0.90) across all 4 bins.\n\nThe first version takes a DataFrame where each row corresponds to a given four-taxon set. The data frame is modified by having an additional another column containing the outlier p-values corresponding to each four-taxon set.\nThe second version takes a DataCF object and modifies it by updating the expected concordance factors stored in that object.\nThe last version (which all others call) assumes that the expected concordance factors in the DataCF object are correctly calculated from the test network.\n\narguments\n\noptbl: when false, the loglik field of net is updated but branch lengths are taken as is. When true, a copy of net is used to conduce the test, with updated branch lengths (in coalescent units) and updated loglik. This network is returned.\nquartetstat = :maxCF: test statistic used to obtain an outlier p-value for each four-taxon set. By default, it is the absolute difference between the observed CF and expected CF of the major resolution (which has the largest CF) if quartetstat=:maxCF, as used in Stenz et al. (2015). The other option is quartetstat=:minpval, in which case the absolute difference between the observed CF and expected CF is calculated for all 3 resolutions, leading to 3 (non-independent) p-values. The outlier p-value for a given four-taxon set is taken as the minimum of these 3 p-values. This option can detect all kinds of departure from the network model for a given four-taxon set, but is not recommended because it is very liberal.\ntest = :onesided: the overall goodness-of-fit test performed on binned frequencies of quartet outlier p-values; see above.\n\noutput\n\np-value of the overall goodness-of-fit test\ntest statistic (z value)\na dictionary of the count of p-values in each of the four category\na vector of two test parameters for the Dirichlet distribution:\nvalue of the concentration parameter α\nvalue of the pseudo likelihood (optimized at α)\na vector of outlier p-values, one for each four-taxon set\nnetwork (first and second versions): net with loglik field updated if optbl is false;  copy of net with optimized branch lengths and loglik if optbl is true\n\nreferences\n\nNWM Stenz, B Larget, DA Baum and C Ané (2015). Exploring tree-like and non-tree-like patterns using genome sequences: An example using the inbreeding plant species Arabidopsis thaliana (L.) Heynh. Systematic Biology, 64(5):809-823. doi: 10.1093/sysbio/syv039\n\nsee also this addendum\n\n\n\n\n\n","category":"method"},{"location":"#QuartetNetworkGoodnessFit.jl","page":"home","title":"QuartetNetworkGoodnessFit.jl","text":"","category":"section"},{"location":"","page":"home","title":"home","text":"Julia package to measure the goodness of fit of a phylogenetic network to data on subsets of 4 tips. It depends on the PhyloNetworks package.","category":"page"},{"location":"","page":"home","title":"home","text":"For a tutorial, see the manual:","category":"page"},{"location":"","page":"home","title":"home","text":"Pages = [\n    \"man/gof.md\",\n    \"man/simulate.md\",\n    \"man/expected_qCFs.md\",\n]\nDepth = 1","category":"page"},{"location":"","page":"home","title":"home","text":"References:","category":"page"},{"location":"","page":"home","title":"home","text":"Ruoyi Cai & Cécile Ané (2021). Assessing the fit of the multi-species network coalescent to multi-locus data. Bioinformatics, 37(5):634-641.\nNoah W. M. Stenz, Bret Larget, David A. Baum, and Cécile Ané (2015). Exploring tree-like and non-tree-like patterns using genome sequences: An example using the inbreeding plant species Arabidopsis thaliana (L.) Heynh. Systematic Biology, 64(5):809-823.\nAddendum describing a modification to the model in the original TICR test.","category":"page"},{"location":"#functions","page":"home","title":"functions","text":"","category":"section"},{"location":"","page":"home","title":"home","text":"Pages = [\"lib/public.md\", \"lib/internals.md\"]\nOrder = [:function]","category":"page"},{"location":"lib/internal/#internal-documentation","page":"internal","title":"internal documentation","text":"","category":"section"},{"location":"lib/internal/","page":"internal","title":"internal","text":"Documentation for QuartetNetworkGoodnessFit's internal functions. Those functions are not exported, but can still be used (like: QuartetNetworkGoodnessFit.foo() for a function named foo()).","category":"page"},{"location":"lib/internal/#index","page":"internal","title":"index","text":"","category":"section"},{"location":"lib/internal/","page":"internal","title":"internal","text":"Pages = [\"internals.md\"]","category":"page"},{"location":"lib/internal/#functions","page":"internal","title":"functions","text":"","category":"section"},{"location":"lib/internal/","page":"internal","title":"internal","text":"Modules = [QuartetNetworkGoodnessFit]\nPublic  = false\nOrder   = [:function]","category":"page"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.dirichlet_max-Tuple{DataCF}","page":"internal","title":"QuartetNetworkGoodnessFit.dirichlet_max","text":"dirichlet_max(dcf::DataCF)\n\nCalculate outlier p-values, one for each four-taxon set, using the maximum concordance factor under a Dirichlet distribution. Used by ticr!.\n\noutput\n\nvector of outlier p-values, one for each 4-taxon set\nvalue of the concentration parameter α\nvalue of the pseudo likelihood (optimized at α)\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.dirichlet_min-Tuple{DataCF}","page":"internal","title":"QuartetNetworkGoodnessFit.dirichlet_min","text":"dirichlet_min(dcf::DataCF)\n\nFirst calculate outlier p-values using each of the three concordance factors for each quartet under a Dirichlet distribution, then take the smallest p-value among the three, as the outlier p-value for each four-taxon set.\n\noutput\n\na vector of outlier p-values, one for each quartet\nvalue of the concentration parameter α\nvalue of the pseudo likelihood (optimized at α)\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.expectedCF_ordered","page":"internal","title":"QuartetNetworkGoodnessFit.expectedCF_ordered","text":"expectedCF_ordered(dcf::DataCF, net::HybridNetwork, suffix=\"\"::AbstractString)\n\nExpected quartet concordance factors in dcf, but ordered as they would be if output by PhyloNetworks.countquartetsintrees. Output:\n\n2-dimentional SharedArray (number of 4-taxon sets x 3). dcf.quartet[i].qnet.expCF[j] for 4-taxon set i and resolution j is stored in row qi and column k if qi is the rank of 4-taxon set i (see PhyloNetworks.quartetrank). This rank depends on how taxa are ordered.\nvector of taxon names, whose order matters. These are tip labels in net with suffix suffix added, then ordered alphabetically, or numerically if taxon names can be parsed as integers.\n\n\n\n\n\n","category":"function"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.multinom_lrt!-Tuple{AbstractVector{Float64}, Vector{Quartet}}","page":"internal","title":"QuartetNetworkGoodnessFit.multinom_lrt!","text":"multinom_lrt!(pval::AbstractVector{Float64}, quartet::Vector{Quartet})\nmultinom_lrt!(pval::AbstractVector{Float64}, obsCF, expCF::AbstractMatrix{Float64})\n\nCalculate outlier p-values (one per four-taxon set) using the likelihood ratio test under a multinomial distribution for the observed concordance factors.\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.multinom_pearson!-Tuple{AbstractVector{Float64}, Vector{Quartet}}","page":"internal","title":"QuartetNetworkGoodnessFit.multinom_pearson!","text":"multinom_pearson!(pval::AbstractVector{Float64}, quartet::Vector{Quartet})\nmultinom_pearson!(pval::AbstractVector{Float64}, obsCF, expCF::AbstractMatrix{Float64})\n\nCalculate outlier p-values (one per four-taxon set) using Pearson's chi-squared statistic under a multinomial distribution for the observed concordance factors.\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.multinom_qlog!-Tuple{AbstractVector{Float64}, Vector{Quartet}}","page":"internal","title":"QuartetNetworkGoodnessFit.multinom_qlog!","text":"multinom_qlog!(pval::AbstractVector{Float64}, quartet::Vector{Quartet})\nmultinom_qlog!(pval::AbstractVector{Float64}, obsCF, expCF::AbstractMatrix{Float64})\n\nCalculate outlier p-values (one per four-taxon set) using the Qlog statistic (Lorenzen, 1995), under a multinomial distribution for the observed concordance factors.\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.network_expectedCF!-Tuple{PhyloNetworks.QuartetT{StaticArraysCore.MVector{3, Float64}}, HybridNetwork, Any, Any, Any}","page":"internal","title":"QuartetNetworkGoodnessFit.network_expectedCF!","text":"network_expectedCF!(quartet::QuartetT, net::HybridNetwork, taxa, taxonnumber,\n        inheritancecorrelation)\n\nUpdate quartet.data to contain the quartet concordance factors expected from the multispecies coalescent along network net for the 4-taxon set taxa[quartet.taxonnumber]. taxa should contain the tip labels in net. quartet.taxonnumber gives the indices in taxa of the 4 taxa of interest. taxonnumber should be a dictionary mapping taxon labels in to their indices in taxa, for easier lookup.\n\nnet is not modified.\n\nFor inheritancecorrelation see network_expectedCF. Its value should be between 0 and 1 (not checked by this internal function).\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.network_expectedCF_4taxa!-Tuple{HybridNetwork, Any, Any}","page":"internal","title":"QuartetNetworkGoodnessFit.network_expectedCF_4taxa!","text":"network_expectedCF_4taxa!(net::HybridNetwork, fourtaxa, inheritancecorrelation)\n\nReturn the quartet concordance factors expected from the multispecies coalescent along network net, where the 3 quartet topologies are ordered following the ordering of taxon names in fourtaxa, that is: if fourtaxa is a,b,c,d, then the concordance factors are listed in this order:\n\n(qCF(ab|cd), qCF(ac|bd), qCF(ad,bc))\n\nAssumptions about net:\n\nhas 4 taxa, and those are the same as fourtaxa\nno degree-2 nodes, except perhaps for the root\nedge lengths are non-missing\nhybrid edge γ's are non-missing\n\nThe network is modified as follows: what's above the LSA is removed, the 2 edges incident to the root are fused (if the root is of degree 2), and external degree-2 blobs are removed. net is then simplified recursively by removing hybrid edges for the recursive calculation of qCFs.\n\nFor inheritancecorrelation see network_expectedCF. Its value should be between 0 and 1 (not checked by this internal function).\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.quarnetGoFtest-Tuple{Vector{Quartet}, Function}","page":"internal","title":"QuartetNetworkGoodnessFit.quarnetGoFtest","text":"quarnetGoFtest(quartet::Vector{Quartet}, outlierp_fun!::Function)\nquarnetGoFtest(outlier_pvalues::AbstractVector)\n\nCalculate an outlier p-value for each quartet according to function outlierp_fun! (or take outlier-values as input: second version) and calculate the z-value to test the null hypothesis that 5% of the p-values are < 0.05, versus the one-sided alternative of more outliers than expected.\n\nSee quarnetGoFtest! for more details.\n\nOutput:\n\nz-value\noutlier p-values (first version only)\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.quarnetGoFtest_simulation-Tuple{HybridNetwork, DataCF, Function, Int64, Int64, Bool, Bool}","page":"internal","title":"QuartetNetworkGoodnessFit.quarnetGoFtest_simulation","text":"quarnetGoFtest_simulation(net::HybridNetwork, dcf::DataCF, outlierp_fun!::Function,\n                          seed::Int, nsim::Int, verbose::Bool, keepfiles::Bool)\n\nSimulate gene trees under the multispecies coalescent model along network net using PhyloCoalSimulations. The quartet concordance factors (CFs) from these simulated gene trees are used as input to outlierp_fun! to categorize each 4-taxon set as an outlier (p-value < 0.05) or not. For each simulated data set, a goodness-of-fit z-value is calculated by comparing the proportion of outlier 4-taxon sets to 0.05. The standard deviation of these z-values (assuming a mean of 0), and the z-values themselves are returned.\n\nUsed by quarnetGoFtest!.\n\nWarning: The quartet CFs expected from net are assumed to be stored in dcf.quartet[i].qnet.expCF. This is not checked.\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.reroot!-Tuple{Any, Any}","page":"internal","title":"QuartetNetworkGoodnessFit.reroot!","text":"reroot!(net, refnet)\n\nReroot net to minimize the hardwired cluster distance between the net (with the new root position) and the reference network refnet. Candidate root positions are limited to internal nodes (excluding leaves) that are compatible with the direction of hybrid edges.\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.ticr_optimalpha-Tuple{DataCF}","page":"internal","title":"QuartetNetworkGoodnessFit.ticr_optimalpha","text":"ticr_optimalpha(dcf::DataCF)\n\nFind the concentration parameter α by maximizing the pseudo-log-likelihood of observed quartet concordance factors. The model assumes a Dirichlet distribution with mean equal to the expected concordance factors calculated from a phylogenetic network (under ILS). These expected CFs are assumed to be already calculated, and stored in dcf.\n\nWhen calculating the pseudo-log-likelihood, this function checks the observed concordance factors for any values equal to zero: they cause a problem because the Dirichlet density is 0 at 0 (for concentration α > 1). Those 0.0 observed CF values are re-set to the minimum of:\n\nthe minimum of all expected concordance factors, and\nthe minimum of all nonzero observed concordance factors.\n\noutput\n\nmaximized pseudo-loglikelihood\nvalue of α where the pseudo-loglikelihood is maximized\nreturn code of the optimization\n\nThe optimization uses NLOpt, with the :LN_BOBYQA method. Optional arguments can tune the optimization differently: nloptmethod, xtol_rel (1e-6 by default), starting α value x_start (1.0 by default).\n\n\n\n\n\n","category":"method"},{"location":"lib/internal/#QuartetNetworkGoodnessFit.ultrametrize!-Tuple{HybridNetwork, Bool}","page":"internal","title":"QuartetNetworkGoodnessFit.ultrametrize!","text":"ultrametrize!(net::HybridNetwork, verbose::Bool)\n\nAssign values to missing branch lengths in net to make the network time-consistent (all paths from the root to a given hybrid node have the same length) and ultrametric (all paths from the root to the tips have the same length), if possible. Warnings are given if it's not possible and if verbose is true.\n\nOutput: true if the modified network is ultrametric, false otherwise.\n\nThe major tree is used to calculate the distance from nodes to the root. If a tree edge has a missing length, this length is changed to the following:\n\n0 if the edge is internal,\nthe smallest value possible to make the network ultrametric if the edge is external.\n\nIt is assumed that hybrid nodes are not leaves, such that external edges are necessarily tree edges. If a hybrid edge has a missing length, this length is changed as follows:\n\nIf both partner hybrid edges lack a length: the shortest lengths are assigned to make the network time-consistent at the hybrid node. In particular, either the major edge or the minor edge is assigned length 0.0.\nOtherwise: the value needed to make the network time-consistent considering based on the partner edge's length if this value is non-negative, and 0 if the ideal value is negative.\n\n\n\n\n\n","category":"method"}]
}
